mongo/client_bulk_write.go

```diff
@@ -0,0 +1,706 @@
+	if err := cur.extractError(); err != nil {
+		err.Raw = raw
+		if mb.writeErrors == nil {
+			mb.writeErrors = make(map[int]WriteError)
+		}
+		mb.writeErrors[idx] = *err
+		return false
+	}
+
+	if mb.result.DeleteResults == nil {
+		mb.result.DeleteResults = make(map[int]ClientDeleteResult)
+	}
+	mb.result.DeleteResults[idx] = ClientDeleteResult{int64(cur.N)}
+
+	return true
+}
+
+func (mb *modelBatches) appendInsertResult(cur *cursorInfo, raw bson.Raw) bool {
+	idx := int(cur.Idx) + mb.offset
+	if err := cur.extractError(); err != nil {
+		err.Raw = raw
+		if mb.writeErrors == nil {
+			mb.writeErrors = make(map[int]WriteError)
+		}
+		mb.writeErrors[idx] = *err
+		return false
+	}
+
+	if mb.result.InsertResults == nil {
+		mb.result.InsertResults = make(map[int]ClientInsertResult)
+	}
+	mb.result.InsertResults[idx] = ClientInsertResult{mb.newIDMap[idx]}
+
+	return true
+}
+
+func (mb *modelBatches) appendUpdateResult(cur *cursorInfo, raw bson.Raw) bool {
+	idx := int(cur.Idx) + mb.offset
+	if err := cur.extractError(); err != nil {
+		err.Raw = raw
+		if mb.writeErrors == nil {
+			mb.writeErrors = make(map[int]WriteError)
+		}
+		mb.writeErrors[idx] = *err
+		return false
+	}
+
+	if mb.result.UpdateResults == nil {
+		mb.result.UpdateResults = make(map[int]ClientUpdateResult)
+	}
+	result := ClientUpdateResult{
+		MatchedCount: int64(cur.N),
+	}
+	if cur.NModified != nil {
+		result.ModifiedCount = int64(*cur.NModified)
+	}
+	if cur.Upserted != nil {
+		result.UpsertedID = cur.Upserted.ID
+	}
+	mb.result.UpdateResults[idx] = result
+
+	return true
+}
+
+type clientInsertDoc struct {
+	namespace int
+	document  interface{}
+
+	sizeLimit int
+}
+
+func (d *clientInsertDoc) marshal(bsonOpts *options.BSONOptions, registry *bson.Registry) (interface{}, bsoncore.Document, error) {
+	uidx, doc := bsoncore.AppendDocumentStart(nil)
+
+	doc = bsoncore.AppendInt32Element(doc, "insert", int32(d.namespace))
+	f, err := marshal(d.document, bsonOpts, registry)
+	if err != nil {
+		return nil, nil, err
+	}
+	if d.sizeLimit > 0 && len(f) > d.sizeLimit {
+		return nil, nil, driver.ErrDocumentTooLarge
+	}
+	var id interface{}
+	f, id, err = ensureID(f, bson.NilObjectID, bsonOpts, registry)
+	if err != nil {
+		return nil, nil, err
+	}
+	doc = bsoncore.AppendDocumentElement(doc, "document", f)
+	doc, err = bsoncore.AppendDocumentEnd(doc, uidx)
+	return id, doc, err
+}
+
+type clientUpdateDoc struct {
+	namespace      int
+	filter         interface{}
+	update         interface{}
+	hint           interface{}
+	arrayFilters   []interface{}
+	collation      *options.Collation
+	upsert         *bool
+	multi          bool
+	checkDollarKey bool
+
+	sizeLimit int
+}
+
+func (d *clientUpdateDoc) marshal(bsonOpts *options.BSONOptions, registry *bson.Registry) (bsoncore.Document, error) {
+	uidx, doc := bsoncore.AppendDocumentStart(nil)
+
+	doc = bsoncore.AppendInt32Element(doc, "update", int32(d.namespace))
+
+	f, err := marshal(d.filter, bsonOpts, registry)
```

Comment:
Setting an empty filter / not setting a filter at all yields vague error messages. Both

```go
updateOneModel := mongo.NewClientUpdateOneModel().SetFilter(bson.D{})
```

Will output `document is nil`. Should we qualify this? E.g.

```go
fmt.Errorf("%w: filter is required", err) // where err comes from marshal
```

This case makes the error message particularly confusing:

```go
updateOneModel := mongo.NewClientUpdateOneModel().SetUpdate(bson.D{{"set", bson.D{{"status", "D"}}}})
```

Reply:
If the filter is nil, should we just default it to empty? The specifications don't cover this AFAICT